### PR TITLE
Use pytorch parallelization with dist.gather and mp.Queue in Estimation stage

### PR DESCRIPTION
**Pull request by drewoldag**
## Code Quality
- [ ] My code follows [the code style of this project](https://rail-hub.readthedocs.io/en/latest/source/contributing.html#naming-conventions)
- [ ] I have written unit tests or justified all instances of `#pragma: no cover`; in the case of a bugfix, a new test that breaks as a result of the bug has been added
- [ ] My code contains relevant comments and necessary documentation for future maintainers; the change is reflected in applicable demos/tutorials (with output cleared!) and added/updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] Any breaking changes, described above, are accompanied by backwards compatibility and deprecation warnings

### CODE REVIEW

File: src/rail/estimation/algos/deepdisc.py [hunk 1]
```diff
@@ -456,36 +508,38 @@ def _process_chunk(self, start_idx, metadata):
...
-        # don't write out if no pdfs are returned from the model
-        if num_pdfs == 0:
-            print("No PDFs returned from the model, skipping this chunk")
-            return
+        # call detectron2's `launch` function to parallelize the inference
+        launch(
+            _do_inference,
+            num_gpus_per_machine=2, #! correct this to be a StageParam
+            num_machines=1, #! correct this to be a StageParam
+            machine_rank=0, #? Need to figure out how to get this, maybe from the comm? or ceci???
```
Comment (OliviaLynn) [inline src/rail/estimation/algos/deepdisc.py:519]: This might just be a copy/paste artifact, but do things go wrong if we use the _process_chunk param `rank` for this? It is familiar to me that ceci can and does keep track of machine ranks, but of course integrating ceci with pytorch is absolutely uncharted territory so I could see this being wonky

Comment (drewoldag) [reply src/rail/estimation/algos/deepdisc.py:542]: Not quite copy/paste, but intentional hardcoding just to match the configuration that I've got for experimenting. I think that you're right, and this should use the param `rank`.

File: src/rail/estimation/algos/deepdisc.py [hunk 2]
```diff
@@ -358,7 +365,62 @@ def finalize(self):
...
+        mapper = RedshiftDictMapperEval(
+            DC2ImageReader(), lambda dataset_dict: dataset_dict["filename"]
+        ).map_data
+
+        loader = d2data.build_detection_test_loader(
+            metadata, mapper=mapper, batch_size=batch_size
+        )
+
+        # this batched version will break up the metadata across GPUs under the hood.
+        true_zs, pdfs, ids = run_batched_match_redshift(loader, predictor, ids=True)
```
Comment (grantmerz) [inline src/rail/estimation/algos/deepdisc.py:385]: It would be useful to make the ids a stage config dict option, although this can be a separate PR

File: src/rail/estimation/algos/deepdisc.py [hunk 3]
```diff
@@ -528,6 +600,7 @@ def _do_chunk_output(self, qp_dstn, start, end, first):
             Whether this is the first time writing to the output file, used to
             initialize the file.
         """
+        print('do_chunk inputs', qp_dstn.npdf, start, end, first)
```
Comment (grantmerz) [inline src/rail/estimation/algos/deepdisc.py:603]: We can also remove this print statement

File: src/rail/estimation/algos/deepdisc.py [hunk 4]
```diff
@@ -496,22 +565,25 @@ def _write_temp_file(self, qp_dstn, start_idx, num_pdfs):
             The qp.ensemble to write to the temporary file
         start_idx : int
             The starting index of the block of images
-        num_pdfs : int
-            The number of pdfs in this block of images
         """
         print("Writing out this temporary ensemble to disk")
```
Comment (grantmerz) [inline src/rail/estimation/algos/deepdisc.py:569]: I think we can remove this print statement

File: src/rail/estimation/algos/deepdisc.py [hunk 5]
```diff
@@ -551,12 +629,16 @@ def finalize(self):
         # open each temporary file, write it's contents to the final file.
         is_first = True
         previous_index = 0
+        print(self._temp_file_meta_tuples)
```
Comment (grantmerz) [inline src/rail/estimation/algos/deepdisc.py:632]: and this one

File: src/rail/estimation/algos/deepdisc.py [hunk 6]
```diff
@@ -551,12 +629,16 @@ def finalize(self):
...
         for meta in self._temp_file_meta_tuples:
             tmp_handle = meta.file_handle
             qp_dstn = tmp_handle.read()
             self._do_chunk_output(qp_dstn, previous_index, previous_index + meta.total_pdfs, is_first)
             previous_index += meta.total_pdfs
             is_first = False
 
+        # finalize the output file
+        self._output_handle.finalize_write()
+
```
Comment (grantmerz) [inline src/rail/estimation/algos/deepdisc.py:642]: Did we ever figure out if we the order of lines 641/644 matter?

**Review by grantmerz — APPROVED** [review]

File: src/rail/estimation/algos/deepdisc.py (hunk 6 above)
Comment (drewoldag) [reply src/rail/estimation/algos/deepdisc.py:642]: Yes, I do believe that this matters, and here we have them in the correct order. 
`self._output_handle.finalize_write()` will close the output file for writing. Then the call to `PipelineStage.finalize(self)` will rename the output file from "inprogress_<blahblah>.hdf5" to "<blahblah>.hdf5".

File: src/rail/estimation/algos/deepdisc.py (hunk 5 above)
Comment (drewoldag) [reply src/rail/estimation/algos/deepdisc.py:632]: Done

File: src/rail/estimation/algos/deepdisc.py (hunk 3 above)
Comment (drewoldag) [reply src/rail/estimation/algos/deepdisc.py:603]: Done

File: src/rail/estimation/algos/deepdisc.py (hunk 4 above)
Comment (drewoldag) [reply src/rail/estimation/algos/deepdisc.py:569]: Done

File: src/rail/estimation/algos/deepdisc.py (hunk 2 above)
Comment (drewoldag) [reply src/rail/estimation/algos/deepdisc.py]: Done. I've added a new boolean stage parameter called `return_ids_with_inference` (the name may not be ideal) and piped it through to _do_inference. 

Since run_batched_match_redshift will return `[]` if ids=False, we can let the logic remain as is, and only introduce conditional logic to control whether or not we try to add `all_ids` to the ancillary data of the qp ensemble.
